pkg/services/ngalert/state/manager.go

```diff
@@ -164,18 +163,53 @@ func (st *Manager) Get(orgID int64, alertRuleUID, stateId string) *State {
 	return st.cache.get(orgID, alertRuleUID, stateId)
 }
 
-// ResetStateByRuleUID deletes all entries in the state manager that match the given rule UID.
-func (st *Manager) ResetStateByRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey) []*State {
+// ResetStateByDeletedRuleUID deletes all entries in the state manager that match the given rule UID.
+func (st *Manager) ResetStateByDeletedRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey) []*State {
+	return st.resetStateByRuleUID(ctx, ruleKey, nil, ngModels.StateReasonDeleted)
+}
+
+// ResetStateByPausedRuleUID deletes all entries in the state manager that match the given rule UID.
+func (st *Manager) ResetStateByPausedRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey, rule *ngModels.AlertRule) []*State {
+	return st.resetStateByRuleUID(ctx, ruleKey, rule, ngModels.StateReasonPaused)
+}
+
+// resetStateByRuleUID removes the rule instances from cache and instanceStore. If reason is ngModels.StateReasonPaused
+// also adds an entry to state history. rule argument must not be nil when the reason is ngModels.StateReasonPaused as
+// it is needed to add the entry to the state history, otherwise it can be nil.
+func (st *Manager) resetStateByRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey, rule *ngModels.AlertRule, reason string) []*State {
 	logger := st.log.New(ruleKey.LogContext()...)
 	logger.Debug("Resetting state of the rule")
+
 	states := st.cache.removeByRuleUID(ruleKey.OrgID, ruleKey.UID)
 	if len(states) > 0 && st.instanceStore != nil {
 		err := st.instanceStore.DeleteAlertInstancesByRule(ctx, ruleKey)
 		if err != nil {
 			logger.Error("Failed to delete states that belong to a rule from database", "error", err)
 		}
 	}
-	logger.Info("Rules state was reset", "states", len(states))
+
+	if reason == ngModels.StateReasonPaused {
+		transitions := make([]StateTransition, 0, len(states))
+		for _, s := range states {
+			oldState := s.State
+			oldReason := s.StateReason
+			now := time.Now()
+			s.SetNormal(reason, s.StartsAt, now)
+			s.LastEvaluationTime = now
+			s.Values = map[string]float64{}
+			transitions = append(transitions, StateTransition{
+				State:               s,
+				PreviousState:       oldState,
+				PreviousStateReason: oldReason,
+			})
+		}
+
+		if st.historian != nil {
+			st.historian.RecordStatesAsync(ctx, rule, transitions)
+		}
+	}
+
+	logger.Info("rules state was reset", "states", len(states))
```

Comment:
According to the Grafana logging guidelines, the message should start from capital letter.